assets/app/view/game/part/town_rect.rb

```diff
@@ -35,63 +35,113 @@ class TownRect < Base
-          gentle: City::EDGE_CITY_REGIONS,
+          0 => [21],
+          0.5 => [13, 21],
+          1 => [13],
+          1.5 => [6, 13],
+          2 => [6],
+          2.5 => [2, 6],
+          3 => [2],
+          3.5 => [2, 10],
+          4 => [10],
+          4.5 => [10, 17],
+          5 => [17],
+          5.5 => [17, 21],
+        }.freeze
+
+        # bias away from top and bottom if possible
+        EDGE_TOWN_REVENUE_REGIONS = {
+          0 => [[23], false],
+          0.5 => [[12], true],
+          1 => [[5], true],
+          1.5 => [[19], false],
+          2 => [[12], false],
+          2.5 => [[5], false],
+          3 => [[0], false],
+          3.5 => [[11], true],
+          4 => [[18], true],
+          4.5 => [[23], true],
+          5 => [[11], false],
+          5.5 => [[18], false],
         }.freeze
 
         # absolute value of angle (in degrees) to tilt the town rectangle relative
         # to a straight line from edge_a to the opposite edge
         RECTANGLE_TILT = {
-          sharp: 10,
-          gentle: 10,
+          sharp: 40,
+          gentle: 15,
         }.freeze
 
         # from center of the hex, how many degrees of offset to apply to
         # positioning the town rectangle; this value will be added/subtracted to
         # (edge_a * 60)
         POSITIONAL_ANGLE = {
-          sharp: 2,
-          gentle: 2,
+          sharp: 12.12,
+          gentle: 6.11,
         }.freeze
 
         DOUBLE_DIT_REVENUE_ANGLES = [10, -130, 130, -10, 50, -50].freeze
         DOUBLE_DIT_REVENUE_REGIONS = City::OO_REVENUE_REGIONS
 
-        # Returns the two edges so that a is the edge to render close to, and
+        # Returns the two edges plus min edge so that a is the edge to render close to, and
         # (a - b).abs is 1, 2 or 3.
-        def normalized_edges
-          @normalized_edges ||=
-            begin
-              edge_a = @tile.preferred_city_town_edges[@town]
-              edge_b = (@town.exits - [edge_a]).first
-              edges = [edge_a, edge_b]
-              edges[edges.index(edges.min)] += 6 if (edge_a - edge_b).abs > 3
-              edges
-            end
-        end
-
-        def min_edge
-          @min_edge ||= normalized_edges.min
+        def self.normalized_edges(edge_a, exits)
+          if edge_a && exits.size == 2
+            edge_b = (exits - [edge_a]).first
+            edges = [edge_a, edge_b]
+            edges[edges.index(edges.min)] += 6 if (edge_a - edge_b).abs > 3
+            edges.append(edges.min)
+          else
+            [edge_a, nil, nil]
+          end
         end
 
         # Returns a symbol of :straight, :gentle or :sharp based on the absolute
         # difference of the normalized edges
-        def track_type
-          @track_type ||=
+        def self.track_type(edges)
+          @@track_type ||= {}
+          @@track_type[edges] ||=
             begin
-              edge_a, edge_b = normalized_edges
+              edge_a, edge_b, = edges
```

Comment:
why is there an extra comma here

Reply:
why is there a nil in the array?

Reply:
could this also be [:sharp, :gentle, :straight[... - 1] ? also maybe make that into a constant